include/zephyr/sys/time_units.h

```diff
@@ -1433,6 +1433,32 @@ static TIME_CONSTEXPR inline uint64_t k_ticks_to_cyc_ceil64(uint64_t t)
 
 #undef TIME_CONSTEXPR
 
+/**
+ * @brief Wait for an expression to return true with a timeout
+ *
+ * Spin on an expression with a timeout and optional delay between iterations
+ *
+ * Commonly needed when waiting on hardware to complete an asynchronous
+ * request to read/write/initialize/reset, but useful for any expression.
+ *
+ * @param expr Truth expression upon which to poll, e.g.: XYZREG & XYZREG_EN
+ * @param timeout Timeout to wait for in microseconds, e.g.: 1000 (1ms)
+ * @param delay_stmt Delay statement to perform each poll iteration
+ *                   e.g.: NULL, k_yield(), k_msleep(1) or k_busy_wait(1)
+ *
+ * @retval expr As a boolean return, if false then it has timed out.
+ */
+#define WAIT_FOR(expr, timeout, delay_stmt)                                                        \
```

Comment:
I'm not really happy with this movement, even though I understand the motivation/details. I wonder if time_units.h could be made independent of util.h by having a few internal macro copies from util.h... or, another solution would be to create a 3rd "common util" header (meant to be internal) that is included by util.h and time_units.h...